plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.hs

```diff
@@ -348,6 +451,67 @@ extKCS k KCS{..} = KCS{ kindOf = kindOf' }
 
 -- ** Type checking
 
+instance Check TypeBuiltinG TermConstantG where
+  check TyByteStringG (TmByteStringG _) = true
+  check TyIntegerG    (TmIntegerG    _) = true
+  check TyStringG     (TmStringG     _) = true
+  check TyBoolG       (TmBoolG       _) = true
+  check TyCharG       (TmCharG       _) = true
+  check TyUnitG       (TmUnitG       _) = true
+  check _             _                 = false
+
+instance Check (TypeG tyname) TermBuiltinG where
+  check (TyFunG (TyBuiltinG TyIntegerG) (TyFunG (TyBuiltinG TyIntegerG) (TyBuiltinG TyIntegerG))) b = case b of
```

Comment:
That's quite a lot of typing... Could we instead infer all the types for built-in functions, construct some data structure out of them and do lookups? All of that automatically. For example, we could have `Map (TypeG tyname) [TermBuiltinG]` and then we'd only need to check that `b` is an element of the list at the type we have at hand.

We're going to change the set of built-in functions, and likely not once, so retyping all of that manually won't be any fun.

Reply:
Would the inferred types for builtins be in PLC syntax?

This is supposed to be a standalone model of the language for testing. On one hand I think it's good to make use of the generic builtin machinery but on the other hand I don't want to just use the PLC implementation for everything as this would defeat the object of the model. I am not sure where to draw the line.

Reply:
> Would the inferred types for builtins be in PLC syntax?

Not necessarily, you can write your own version [`typeSchemeToType`](https://github.com/input-output-hk/plutus/blob/e337cac32fdd5b8e68e2b7a225f1feb2c9622658/plutus-core/src/Language/PlutusCore/Constant/Function.hs#L23) that returns a `TypeG` instead.

> This is supposed to be a standalone model of the language for testing. On one hand I think it's good to make use of the generic builtin machinery but on the other hand I don't want to just use the PLC implementation for everything as this would defeat the object of the model.

We have golden tests for types of built-in functions. Plus, those types are not specified manually anywhere, they are [derived automatically](https://github.com/input-output-hk/plutus/blob/e337cac32fdd5b8e68e2b7a225f1feb2c9622658/plutus-core/src/Language/PlutusCore/Builtins.hs#L134), so they can't be misspecified (but there's a possibility of misderiving them). Plus, we kind check all built-in functions before running the type checker. I think, the `TypeScheme` of a built-in function is a pretty trusted thing that we check in different places, so we don't need to check it in one more place. So I'd argue NEAT can be a standalone model of the language modulo builtins, but YMMV. It'd just be very convenient to treat it as such.

Reply:
I'm looking at `TypeScheme`

Reply:
I have spent some time trying things out without success yet - one thing I noticed is that `TypeScheme` contains constraints about 
 classes `KnownType` and `KnownKind` which are specific to the PLC syntax which worries me. I was expecting `TypeScheme` to be independent of the syntax.

Reply:
I tried this (with a polymorphi version of TypeG which I thought might help but doesn't):
```
data TypeG' uni tyname
  = TyVarG' tyname
  | TyFunG' (TypeG' uni tyname) (TypeG' uni tyname)
  | TyIFixG' (TypeG' uni tyname) (Kind ()) (TypeG' uni tyname)
  | TyForallG' (Kind ()) (TypeG' uni (S tyname))
  | TyBuiltinG' (Some (TypeIn uni))
  | TyLamG' (TypeG' uni (S tyname))
  | TyAppG' (TypeG' uni tyname) (TypeG' uni tyname) (Kind ())
  deriving (Typeable, Eq, Show, Functor)


mkTyBuiltinG
    :: forall a uni tyname. uni `Includes` a
    => TypeG' uni tyname
mkTyBuiltinG = TyBuiltinG' . Some . TypeIn $ knownUni @uni @a

toBuiltinTypeAstG :: uni `Includes` a => proxy a -> TypeG' uni tyname
toBuiltinTypeAstG (_ :: proxy a) = mkTyBuiltinG @a

class KnownTypeAstG uni (a :: k) where
    -- | The type representing @a@ used on the PLC side.
    toTypeAstG :: proxy a -> TypeG' uni tyname

typeSchemeToTypeG :: TypeScheme term args res -> TypeG' (UniOf term) tyname'
typeSchemeToTypeG (TypeSchemeResult pR)      = toBuiltinTypeAstG pR
typeSchemeToTypeG (TypeSchemeArrow pA schB)  =
  TyFunG' (toBuiltinTypeAstG pA) $ typeSchemeToTypeG schB
typeSchemeToTypeG (TypeSchemeAll proxy schK) = case proxy of
    (_ :: Proxy '(text, uniq, kind)) ->
      TyForallG' (knownKind $ Proxy @kind) $ typeSchemeToTypeG (schK Proxy)

```

I get the following type error:
```
generators/Language/PlutusCore/Generators/NEAT/Type.hs:629:48: error:
    • Could not deduce (Includes (UniOf term) res)
        arising from a use of ‘toBuiltinTypeAstG’
      from the context: (args ~ '[], KnownType term res)
        bound by a pattern with constructor:
                   TypeSchemeResult :: forall term res.
                                       KnownType term res =>
                                       Proxy res -> TypeScheme term '[] res,
                 in an equation for ‘typeSchemeToTypeG’
        at generators/Language/PlutusCore/Generators/NEAT/Type.hs:629:20-38
    • In the expression: toBuiltinTypeAstG pR
      In an equation for ‘typeSchemeToTypeG’:
          typeSchemeToTypeG (TypeSchemeResult pR) = toBuiltinTypeAstG pR
    |
629 | typeSchemeToTypeG (TypeSchemeResult pR)      = toBuiltinTypeAstG pR
    |                                                ^^^^^^^^^^^^^^^^^^^^

generators/Language/PlutusCore/Generators/NEAT/Type.hs:631:12: error:
    • Could not deduce (Includes (UniOf term) arg)
        arising from a use of ‘toBuiltinTypeAstG’
      from the context: (args ~ (arg : args1), KnownType term arg)
        bound by a pattern with constructor:
                   TypeSchemeArrow :: forall term arg (args1 :: [*]) res.
                                      KnownType term arg =>
                                      Proxy arg
                                      -> TypeScheme term args1 res
                                      -> TypeScheme term (arg : args1) res,
                 in an equation for ‘typeSchemeToTypeG’
        at generators/Language/PlutusCore/Generators/NEAT/Type.hs:630:20-42
    • In the first argument of ‘TyFunG'’, namely
        ‘(toBuiltinTypeAstG pA)’
      In the expression: TyFunG' (toBuiltinTypeAstG pA)
      In the expression:
        TyFunG' (toBuiltinTypeAstG pA) $ typeSchemeToTypeG schB
    |
631 |   TyFunG' (toBuiltinTypeAstG pA) $ typeSchemeToTypeG schB
    |            ^^^^^^^^^^^^^^^^^^^^

```

I don't want to waste your time solving my type errors for me but I am a bit stumped as to how to get ahold of `Includes (UniOf term) res` and `Includes (UniOf term) arg`.

Reply:
ah, I found `IncludesAll` which solves one of my problems.

Reply:
> I have spent some time trying things out without success yet - one thing I noticed is that `TypeScheme` contains constraints about classes `KnownType` and `KnownKind` which are specific to the PLC syntax which worries me. I was expecting `TypeScheme` to be independent of the syntax.

`KnownType` being inside of `TypeScheme` is very annoying indeed, it's been bugging me for quite a while. We couldn't get rid of it previously, but it might in fact be possible now. As for `KnownKind`, it's used to reflect Haskell kinds and I just immediately convert those to Plutus ones, because we didn't need any additional flexibility there. We could introduce a singleton for Haskell kinds and use it in `KnownKind` instead of going directly to Plutus kinds.

Your points make a lot of sense and I'd be willing to try generalizing the extensible builtins machinery a bit further to make it less syntax-specific. How about merging this PR as is and I'll try to generalize extensible builtins and make NEAT work with universes once I'm done with the first version of HOAS, which should happen this week?

Reply:
Thanks Roman that's very generous of you. 

I got a little further and the below code type checks:
```
data TypeG' uni tyname
  = TyVarG' tyname
  | TyFunG' (TypeG' uni tyname) (TypeG' uni tyname)
  | TyIFixG' (TypeG' uni tyname) (Kind ()) (TypeG' uni tyname)
  | TyForallG' (Kind ()) (TypeG' uni (S tyname))
  | TyBuiltinG' (Some (TypeIn uni))
  | TyLamG' (TypeG' uni (S tyname))
  | TyAppG' (TypeG' uni tyname) (TypeG' uni tyname) (Kind ())
  deriving (Typeable, Eq, Show, Functor)

_mkTyBuiltinG
    :: forall a uni tyname. uni `Includes` a
    => TypeG' uni tyname
_mkTyBuiltinG = TyBuiltinG' . Some . TypeIn $ knownUni @uni @a

_toBuiltinTypeAstG :: uni `Includes` a => proxy a -> TypeG' uni tyname
_toBuiltinTypeAstG (_ :: proxy a) = _mkTyBuiltinG @a

_typeSchemeToTypeG :: (UniOf term `Includes` res, UniOf term `IncludesAll` args)
  => TypeScheme term args res -> TypeG' (UniOf term) tyname'
_typeSchemeToTypeG (TypeSchemeResult pR)      = _toBuiltinTypeAstG pR
_typeSchemeToTypeG (TypeSchemeArrow pA schB)  =
  TyFunG' (_toBuiltinTypeAstG pA) $ _typeSchemeToTypeG schB
_typeSchemeToTypeG (TypeSchemeAll proxy schK) = case proxy of
    (_ :: Proxy '(text, uniq, kind)) ->
      TyForallG' (knownKind $ Proxy @kind) $ _typeSchemeToTypeG (schK Proxy)
```

I think it would be a good idea to split off the universe stuff into a separate PR. I think removing `KnowKind` from `TypeScheme` would be cleaner. Your suggestion sounds good to me! I am also happy to return to this if you don't have time.